lib/react_client.dart

```diff
@@ -135,69 +162,371 @@ class ReactDartComponentFactoryProxy<TComponent extends Component> extends React
 
 /// Creates ReactJS [Component2] instances for Dart components.
 class ReactDartComponentFactoryProxy2<TComponent extends Component2> extends ReactComponentFactoryProxy
+    with JsBackedMapComponentFactoryMixin
     implements ReactDartComponentFactoryProxy {
   /// The ReactJS class used as the type for all [ReactElement]s built by
   /// this factory.
+  @override
   final ReactClass reactClass;
 
   /// The JS component factory used by this factory to build [ReactElement]s.
+  ///
+  /// Deprecated: Use [React.createElement()] instead and pass in [type] as
+  /// the first argument, followed by `props` and `children`.
+  ///
+  /// Before:
+  /// ```
+  ///   YourFactoryProxy.reactComponentFactory(props, children);
+  /// ```
+  ///
+  /// After:
+  /// ```
+  ///   React.createElement(YourFactoryProxy.type, props, children);
+  /// ```
+  @override
+  @Deprecated('6.0.0')
   final ReactJsComponentFactory reactComponentFactory;
 
+  @override
   final Map defaultProps;
 
   ReactDartComponentFactoryProxy2(ReactClass reactClass)
       : this.reactClass = reactClass,
         this.reactComponentFactory = React.createFactory(reactClass),
-        this.defaultProps = new JsBackedMap.fromJs(reactClass.defaultProps);
+        this.defaultProps = JsBackedMap.fromJs(reactClass.defaultProps);
 
+  @override
   ReactClass get type => reactClass;
 
+  /// Returns a JavaScript version of the specified [props], preprocessed for consumption by ReactJS and prepared for
+  /// consumption by the [react] library internals.
+  static JsMap generateExtendedJsProps(Map props) => JsBackedMapComponentFactoryMixin.generateExtendedJsProps(props);
+}
+
+/// Shared component factory proxy [build] method for components that utilize [JsBackedMap]s.
+mixin JsBackedMapComponentFactoryMixin on ReactComponentFactoryProxy {
+  @override
   ReactElement build(Map props, [List childrenArgs = const []]) {
-    // TODO 3.1.0-wip if we don't pass in a list into React, we don't get a list back in Dart...
+    var children = _generateChildren(childrenArgs, shouldAlwaysBeList: true);
+    return React.createElement(type, JsBackedMapComponentFactoryMixin.generateExtendedJsProps(props), children);
+  }
 
-    List children;
-    if (childrenArgs.isEmpty) {
-      children = childrenArgs;
-    } else if (childrenArgs.length == 1) {
-      final singleChild = listifyChildren(childrenArgs[0]);
-      if (singleChild is List) {
-        children = singleChild;
+  static JsMap generateExtendedJsProps(Map props) => _generateJsProps(props);
+}
+
+JsMap _generateJsProps(Map props,
+    {bool convertEventHandlers = false,
```

Comment:
These arguments don't seem to be used at all; are they needed?

Reply:
https://github.com/cleandart/react-dart/pull/221#discussion_r342769366